C++ Reference Pages

hhenry01 opened the pull request:
Closes #37

hhenry01 committed:
Initial C++ Reference Pages

hhenry01 committed:
Fix linting

patrick-5546 committed:
Add back border color

patrick-5546 commented on docs/reference/cpp/index.md:
````diff
@@ -0,0 +1,172 @@
+
+C++ introduces namespaces.
+
+```C++
+// A.h
+namespace a {
+float x;
+int bar(void);
+}  // namespace a
+
+// B.h
+namespace b {
+int bar(void);
+}  // namespace b
+
+// ighxy.cpp
+#include "A.h"
+#include "B.h"
+
+int main(void) {
+    int a = a::bar();
+    int b = b::bar();
+    float xa = a::x;
+    float xb = b::x;
+    ...
+}
+```
````
I think this could be tabbed content: https://squidfunk.github.io/mkdocs-material/reference/content-tabs/#grouping-code-blocks

Another option would be to use titled code blocks: https://squidfunk.github.io/mkdocs-material/reference/code-blocks/#adding-a-title

patrick-5546 commented on docs/reference/cpp/index.md:
````diff
@@ -0,0 +1,172 @@
+#define PI 3.14F
+#define AREA_OF_CIRCLE(radius) ((PI) * (radius) * (radius))
+
+int main(void) {
+    // Before precompile
+    float area = AREA_OF_CIRCLE(2.5F);
+    // After precompile
+    float area = ((3.14F) * (2.5F) * (2.5F));
+    ...
+}
+```
````
This could be tabbed content as well

patrick-5546 reviewed (commented):
Left some comments to fully take advantage of the features that Material for MkDocs provides. Can't say much about the content itself, will add @DFriend01 as a reviewer for that

hhenry01 committed:
Update references

hhenry01 commented:
I changed some of the formatting to take advantage of the mkdocs-material features better. I also added an extra section on C++ arrays and added a Googletest example in the tools tab. I feel like the other tools are either too complicated or too specific to a one-time setup that I need to perform for examples to be put in.

patrick-5546 reviewed (dismissed):
From an infrastructure and organizational standpoint it LGTM

patrick-5546 commented:
Some thoughts I had after the review
- C vs C++ code blocks might be better shown side by side using tables rather than in seperate tabs so that it is easier to compare. What do you think?
- If you feel like something interrupts the flow of the documentation (e.g., long code block), you can put it in a collapsible admonition: https://squidfunk.github.io/mkdocs-material/reference/admonitions/#collapsible-blocks

DFriend01 reviewed (changes requested):
Ignore this comment.

DFriend01 commented on docs/reference/cpp/index.md:
````diff
@@ -0,0 +1,254 @@
+        ...
+    }
+    ```
+
````
I think it might be prudent to introduce `using namespace` if one wanted to avoid writing, for example, `std::cout` for printing every time. Up to you how you might introduce this or if at all.

hhenry01 replied:
Sounds good, though `using namespace std` is very bad practice so that shouldn't be the example.

DFriend01 commented on docs/reference/cpp/index.md:
````diff
@@ -0,0 +1,254 @@
+        ...
+    }
+    ```
+
````
It might be worth it to mention that AREA_OF_CIRCLE is a macro with arguments. You can link them to [this](https://www.cs.yale.edu/homes/aspnes/pinewiki/C(2f)Macros.html) resource that explains C macros nicely.

DFriend01 commented on docs/reference/cpp/tools.md:
```diff
@@ -0,0 +1,115 @@
+# Tools
+
+A lot goes into making a well structured C++ project, much more than any one team should have to do.
+
```
Do you use debugging tools like GDB? It might be worth mentioning them here if you use them often.

hhenry01 replied:
Good point, I'll likely need to elaborate on the specific quirks of GDB with VSCode.

DFriend01 reviewed (changes requested):
I left a few comments. Reply to them if you have any questions.

I have some ideas for more content that we may want to add with respect to some conventions that we want to tell our new recruits to follow. This may be things like linting, documentation styles, coding style, etc.

Take a look at the [conventions page for Python](https://ubcsailbot.github.io/docs/reference/python/conventions/) for some ideas.

DFriend01 commented on docs/reference/cpp/index.md:
```diff
@@ -0,0 +1,254 @@
+Feel free to add other resources other than the ones listed above if you find any that you like!
+
+## Differences Between C and C++
+
```
This can belong on its own page in another markdown file. Then, add it to the subtree under C++.

hhenry01 committed:
Resolve PR discussion comments

hhenry01 commented:
Side by side code blocks look to crowded so I decided against it.

DFriend01 committed:
Moved Getting Started page as separate page and reordered C++ pages

hhenry01 committed:
Merge branch 'main' into C++-Page
